File: 113_PathSumII/113_PathSumII.cpp
class Solution {
public:
    // iterative solution, post-order travel
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        stack<TreeNode*> dfs;
        TreeNode *pre = NULL, *cur = root, *top = NULL;
        vector<int> v;
        vector<vector<int>> res;
        int val = 0;
        while( cur || !dfs.empty() ){
            if(cur){
                dfs.push(cur);
                v.push_back(cur -> val);
                val += cur -> val;
                if( !cur -> left && !cur -> right && val == sum)
                    res.push_back(v);
                cur = cur -> left;
            }else{
                TreeNode *top = dfs.top();
                if( top -> right == pre || !top -> right ){
                    dfs.pop();
                    v.pop_back();
                    val -= top -> val;
                    pre = top;
                }else{
                    cur = top -> right;
                }
            }
        }
        return res;
        
    }
};

File: 113_PathSumII/113_PathSumII.py
class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root: return []
        DFS = [(root, 0, root.val)]
        l=[]
        res=[]
        while DFS:
            node, depth, s = DFS.pop()
            if(depth+1 > len(l)): l.append(node.val)
            else: l[depth] = node.val
            if(not node.left and not node.right and s==sum):
                res.append(l[:depth+1])
            if(node.right):
                DFS.append((node.right, depth+1, s+node.right.val))
            if(node.left):
                DFS.append((node.left, depth+1, s+node.left.val))
        return res
